### DISCUSSION TITLE
Importing Python Files During Runtime?

### DISCUSSION THREAD

**Discussion by Rondop** [pyinstaller/pyinstaller#5546] [Help] [answered]
Hello there :-)

I wrote a small Python command-line tool that some people in the office can use. The problem is that most are not Python programmers, and if I wish to distribute the tool using `pip` - it means all of them need Python installed. This tool consume external generated Python code (Protobuf). So for example the user will state to the tool he wish to download version `1.2.1` of the mentioned generated Python code, and the tool will consume those on future commands and do manipulations on top of those. So the generated Python code can't really get bundled. The problem here, is that the input of the tool, is a generate python code that needed to be loaded in real-time.

I'm using `import_module` to import those Python files during run-time, but if I understand correctly, `import_module` is being intercepted by `PyInstaller`. Because those generate Python files were not there during the creation of that exe, there's not `pyc` files for those and the tool fails to work. 

Is there a way to tell `PyInstaller` that some import are runtime imports? or that's break the philosophy of a single tool that bundles all Python code in it?

Thank you!

**Comment by bwoodsend** [comment, marked as answer]
PyInstaller-ed Python still contains everything it needs to run regular external Python code. As long as your code is importable (its parent folder is added to `sys.path`) you should be able to load it with regular `import`.

**Reply by Rondop** [reply]
Thank you for the quick answer bwoodsend. 

In that case, I think I know what wrong. Some of the generated Protobuf code is being supplied by Google (by a package called `protobuf`). However, this package doesn't include all the Google Protobufs (like annotations, api etc). So a common use is to append the extra Python generated code into the google package like so:

```
google.__path__.append(path_to_extra_google_protobufs).
```

I suspect that this error:

```
from google.protobuf import struct_pb2 as google_dot_protobuf_dot_struct__pb2
ImportError: cannot import name 'struct_pb2' from 'google.protobuf' (/Users/rondop/repositories/test_pyinstaller/dist/tool/google/protobuf/__init__.pyc)
```

Is because the binary doesn't bundle the google base code in it. I will see if I can fixing it by manually stating the package name in `hiddenimports`.

Thanks you for tipping me into the right direction! :)

Edit: Confirms it works. Thank you again!
